impeller/geometry/rect.h

```diff
@@ -69,76 +166,114 @@ struct TRect {
+           bottom_ == r.bottom_;
   }
 
   [[nodiscard]] constexpr TRect Scale(Type scale) const {
-    return TRect({origin.x * scale, origin.y * scale},
-                 {size.width * scale, size.height * scale});
+    return TRect(left_ * scale,   //
+                 top_ * scale,    //
+                 right_ * scale,  //
+                 bottom_ * scale);
   }
 
   [[nodiscard]] constexpr TRect Scale(Type scale_x, Type scale_y) const {
-    return TRect({origin.x * scale_x, origin.y * scale_y},
-                 {size.width * scale_x, size.height * scale_y});
+    return TRect(left_ * scale_x,   //
+                 top_ * scale_y,    //
+                 right_ * scale_x,  //
+                 bottom_ * scale_y);
   }
 
   [[nodiscard]] constexpr TRect Scale(TPoint<T> scale) const {
-    return TRect({origin.x * scale.x, origin.y * scale.y},
-                 {size.width * scale.x, size.height * scale.y});
+    return Scale(scale.x, scale.y);
   }
 
   [[nodiscard]] constexpr TRect Scale(TSize<T> scale) const {
-    return Scale(TPoint<T>(scale));
+    return Scale(scale.width, scale.height);
   }
 
+  /// @brief  Returns true iff the provided point |p| is inside the
+  ///         helf-open interior of this rectangle.
+  ///
+  ///         For purposes of containment, a rectangle contains points
+  ///         along the top and left edges but not points along the
+  ///         right and bottom edges so that a point is only ever
+  ///         considered inside one of two abutting rectangles.
   [[nodiscard]] constexpr bool Contains(const TPoint<Type>& p) const {
-    return p.x >= GetLeft() && p.x < GetRight() && p.y >= GetTop() &&
-           p.y < GetBottom();
+    return !this->IsEmpty() &&  //
+           p.x >= left_ &&      //
+           p.y >= top_ &&       //
+           p.x < right_ &&      //
+           p.y < bottom_;
   }
 
+  /// @brief  Returns true iff this rectangle is not empty and it also
+  ///         contains every point considered inside the provided
+  ///         rectangle |o| (as determined by |Contains(TPoint)|).
+  ///
+  ///         This is similar to a definition where the result is true iff
+  ///         the union of the two rectangles is equal to this rectangle,
+  ///         ignoring precision issues with performing those operations
+  ///         and assuming that empty rectangles are never equal.
+  ///
+  ///         An empty rectangle can contain no other rectangle.
+  ///
+  ///         An empty rectangle is, however, contained within any
+  ///         other non-empy rectangle as the set of points it contains
+  ///         is an empty set and so there are no points to fail the
+  ///         containment criteria.
   [[nodiscard]] constexpr bool Contains(const TRect& o) const {
-    return o.GetLeft() >= GetLeft() && o.GetTop() >= GetTop() &&
-           o.GetRight() <= GetRight() && o.GetBottom() <= GetBottom();
+    // Is this->IsEmpty test needed here? Write tests and then evaluate if
```

Comment:
Add a TODO for this?

Reply:
Actually, it ended up being needed after I adjusted the semantics so it's just obsolete now.

Reply:
Done